services/provider/server/consumer.go

```diff
@@ -0,0 +1,158 @@
+	consumers := &ocsv1alpha1.StorageConsumerList{}
+	err := client.List(ctx, consumers)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list storage consumers. %v", err)
+	}
+
+	ticketStore := map[string]string{}
+	uidStore := map[string]string{}
+
+	for _, consumer := range consumers.Items {
+		uid := string(consumer.UID)
+		if _, ok := uidStore[uid]; !ok {
+			uidStore[uid] = consumer.Name
+		}
+
+		if ticket, ok := consumer.GetAnnotations()[TicketAnnotation]; ok {
+			if _, ok := ticketStore[ticket]; !ok {
+				ticketStore[ticket] = consumer.Name
+			}
+		}
+	}
+
+	return &ocsConsumerManager{
+		client:      client,
+		ticketStore: ticketStore,
+		uidStore:    uidStore,
+	}, nil
+}
+
+// Create creates a new storageConsumer resource, updates the consumer cache and returns the storageConsumer UUID
+func (c *ocsConsumerManager) Create(ctx context.Context, name, namespace, ticket string, capacity resource.Quantity) (string, error) {
+	consumerObj := &ocsv1alpha1.StorageConsumer{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+			Annotations: map[string]string{
+				TicketAnnotation: ticket,
+			},
+		},
+		Spec: ocsv1alpha1.StorageConsumerSpec{
+			Capacity: capacity,
+		},
+	}
+
+	err := c.client.Create(ctx, consumerObj)
+	if err != nil {
+		if errors.IsAlreadyExists(err) {
+			return "", err
+		}
+		return "", fmt.Errorf("failed to create storageConsumer resource %q. %v", consumerObj.Name, err)
+	}
+
+	existingConsumer := &ocsv1alpha1.StorageConsumer{}
+	err = c.client.Get(ctx, types.NamespacedName{Name: consumerObj.Name, Namespace: consumerObj.Namespace}, existingConsumer)
+	if err != nil {
+		return "", fmt.Errorf("failed to get UUID from the storageConsumer resource %q. %v", consumerObj.Name, err)
+	}
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	consumerID := string(existingConsumer.UID)
+	if _, ok := c.uidStore[consumerID]; !ok {
+		c.uidStore[consumerID] = existingConsumer.Name
+	}
+	if _, ok := c.ticketStore[ticket]; !ok {
+		c.ticketStore[ticket] = existingConsumer.Name
+	}
+
+	return consumerID, nil
+}
+
+// Delete deletes the storageConsumer resource using UID and updates the consumer cache
+func (c *ocsConsumerManager) Delete(ctx context.Context, namespace, id string) error {
+	c.mutex.RLock()
+	consumerName, ok := c.uidStore[id]
+	if !ok {
+		klog.Warningf("no storageConsumer found with UUID %q", id)
+		c.mutex.RUnlock()
+		return nil
+	}
+	c.mutex.RUnlock()
+
+	consumerObj := &ocsv1alpha1.StorageConsumer{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      consumerName,
+			Namespace: namespace,
+		},
+	}
+	if err := c.client.Delete(ctx, consumerObj); err != nil {
+		if errors.IsNotFound(err) {
+			klog.Warningf("storageConsumer %q not found.", consumerObj.Name)
+			return nil
+		}
+		return fmt.Errorf("failed to delete storageConsumer %q. %v", consumerName, err)
+	}
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
```

Comment:
You would like to remove the lock as soon as possible after the index update, do not defer the operation.
You do not need to handle erroneous states because `delete(map, value)` will not fail if the key does not exist.

Reply:
done.